### PR TITLE
IC-1295 submit attendance

### PR DESCRIPTION
**Pull request by rgforsyth**
## What does this pull request do?

Updated the appointment db table & adds a new endpoint to enable attendance details to be recorded.

## What is the intent behind these changes?

Allows appointment attendance details to be recorded.

### CODE REVIEW

**Comment by tomsmyers** [comment]
small note: please format JIRA tickets like `IC-1234` in the PR titles so they get linked properly. cheers.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/AppointmentsController.kt [hunk 1]
```diff
@@ -83,4 +84,17 @@ class AppointmentsController(
     val actionPlanAppointment = appointmentsService.getAppointment(actionPlanId, sessionNumber)
     return ActionPlanAppointmentDTO.from(actionPlanAppointment)
   }
+
+  @PostMapping("/action-plan/{id}/appointment/{sessionNumber}/record-attendance")
```
Comment (Gweaton) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/AppointmentsController.kt:88]: I'm writing the contracts, and just wondering if this should be the same as the `PATCH` endpoint that we hit to change session duration (eg /action-plan/345059d4-1697-467b-8914-fedec9957279/appointment/2) if we're updating the same object?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/dto/ActionPlanAppointmentsDTO.kt [hunk 2]
```diff
@@ -20,13 +21,19 @@ data class UpdateAppointmentDTO(
   override val durationInMinutes: Int?,
 ) : BaseAppointmentDTO(appointmentTime, durationInMinutes)
 
+data class UpdateAppointmentAttendanceDTO(
+  val attended: SessionAttendance,
+  val additionalAttendanceInformation: String?
+)
```
Comment (mahalrmoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/dto/ActionPlanAppointmentsDTO.kt:27]: Thats a big name ;-) but it'd part of the contract

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/AppointmentsService.kt [hunk 3]
```diff
@@ -54,26 +55,44 @@ class AppointmentsService(
     return actionPlanAppointmentRepository.save(appointment)
   }
 
-  fun getAppointments(actionPlanId: UUID): List<ActionPlanAppointment> {
+  fun updateAppointmentWithAttendance(
```
Comment (mahalrmoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/AppointmentsService.kt:58]: Should the business tier not also reflect the business action being done, i.e. recordAttendance

File: src/main/resources/db/migration/V1_35__appointment_attendance.sql [hunk 4]
```diff
@@ -0,0 +1,3 @@
+alter table action_plan_appointment
```
Comment (mahalrmoj) [inline src/main/resources/db/migration/V1_35__appointment_attendance.sql:1]: as an enum should we have a constraint with valid values?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/AppointmentsService.kt (hunk 3 above)
Comment (rgforsyth) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/AppointmentsService.kt:58]: It probably should. Will change

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/jpa/entity/ActionPlanAppointment.kt [hunk 5]
```diff
@@ -16,6 +19,9 @@ data class ActionPlanAppointment(
 
   // Attributes
   @NotNull val sessionNumber: Int,
+  @Enumerated(EnumType.STRING)
+  var attended: SessionAttendance? = null,
+  var additionalAttendanceInformation: String? = null,
```
Comment (sldblog) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/jpa/entity/ActionPlanAppointment.kt:24]: 💭 Providers have 1 working day (24 hours?) to mark attendance after an elapsed appointment.
Do we capture the time this information is submitted somewhere?

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/AppointmentsControllerTest.kt [hunk 6]
```diff
@@ -109,4 +112,29 @@ internal class AppointmentsControllerTest {
     assertThat(appointmentsResponse.size).isEqualTo(1)
     assertThat(appointmentsResponse.first()).isEqualTo(ActionPlanAppointmentDTO.from(actionPlanAppointment))
   }
+
+  @Test
+  fun `updates appointment with attendance details`() {
```
Comment (sldblog) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/AppointmentsControllerTest.kt:117]: 💭 Should we mention some sort of assertion in the test title, about returning the now "attendance recorded" appointment?

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/AppointmentsServiceTest.kt [hunk 7]
```diff
@@ -209,4 +227,45 @@ internal class AppointmentsServiceTest {
...
+
+    verify(actionPlanAppointmentRepository).save(argumentCaptor.capture())
+    assertThat(argumentCaptor.firstValue.attended).isEqualTo(sessionAttendance)
+    assertThat(argumentCaptor.firstValue.additionalAttendanceInformation).isEqualTo(additionalInformation)
+    assertThat(argumentCaptor.firstValue.attendanceSubmittedAt).isNotNull
+    assertThat(savedAppointment).isNotNull
+  }
+
+  @Test
+  fun `update appointment with attendance - no appointment found`() {
```
Comment (sldblog) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/AppointmentsServiceTest.kt:257]: 💭 Should we add the expectation into the title, like "updating a nonexistent appointment raises an exception"?

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/AppointmentsServiceTest.kt [hunk 8]
```diff
@@ -209,4 +227,45 @@ internal class AppointmentsServiceTest {
...
+      .thenReturn(existingAppointment)
+    whenever(actionPlanAppointmentRepository.save(any())).thenReturn(existingAppointment)
+
+    val savedAppointment = appointmentsService.recordAttendance(appointmentId, 1, sessionAttendance, additionalInformation)
+    val argumentCaptor: ArgumentCaptor<ActionPlanAppointment> = ArgumentCaptor.forClass(ActionPlanAppointment::class.java)
+
+    verify(actionPlanAppointmentRepository).save(argumentCaptor.capture())
+    assertThat(argumentCaptor.firstValue.attended).isEqualTo(sessionAttendance)
+    assertThat(argumentCaptor.firstValue.additionalAttendanceInformation).isEqualTo(additionalInformation)
+    assertThat(argumentCaptor.firstValue.attendanceSubmittedAt).isNotNull
```
Comment (sldblog) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/AppointmentsServiceTest.kt:252]: 💭 Should we assert here that the submission time corresponds to the current time?

**Review by sldblog — COMMENTED** [review]
LGTM! Random, non-blocking comments

**Pull request merged**